substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -0,0 +1,1591 @@
+		);
+
+		let start = Instant::now();
+		let watched_xts = self.register_listeners(&mut view).await;
+		let duration = start.elapsed();
+		log::debug!(target: LOG_TARGET, "register_listeners: at {at:?} took {duration:?}");
+
+		// 2. Handle transactions from the tree route. Pruning transactions from the view first
+		// will make some space for mempool transactions in case we are at the view's limits.
+		let start = Instant::now();
+		self.update_view_with_fork(&view, tree_route, at.clone()).await;
+		let duration = start.elapsed();
+		log::debug!(target: LOG_TARGET, "update_view_with_fork: at {at:?} took {duration:?}");
+
+		// 3. Finally, submit transactions from the mempool.
+		let start = Instant::now();
+		self.update_view_with_mempool(&mut view, watched_xts).await;
+		let duration = start.elapsed();
+		log::debug!(target: LOG_TARGET, "update_view_with_mempool: at {at:?} took {duration:?}");
+
+		let view = Arc::from(view);
+		self.view_store.insert_new_view(view.clone(), tree_route).await;
+		Some(view)
+	}
+
+	/// Returns the list of xts included in all block ancestors, excluding the block itself.
+	///
+	/// For the following chain `F<-B1<-B2<-B3` xts from `F,B1,B2` will be returned.
+	async fn extrinsics_included_since_finalized(&self, at: Block::Hash) -> HashSet<TxHash<Self>> {
+		let start = Instant::now();
+		let recent_finalized_block = self.enactment_state.lock().recent_finalized_block();
+
+		let Ok(tree_route) = self.api.tree_route(recent_finalized_block, at) else {
+			return Default::default()
+		};
+
+		let api = self.api.clone();
+		let mut all_extrinsics = HashSet::new();
+
+		for h in tree_route.enacted().iter().rev().skip(1) {
+			api.block_body(h.hash)
+				.await
+				.unwrap_or_else(|e| {
+					log::warn!(target: LOG_TARGET, "Compute ready light transactions: error request: {}", e);
+					None
+				})
+				.unwrap_or_default()
+				.into_iter()
+				.map(|t| self.hash_of(&t))
+				.for_each(|tx_hash| {
+					all_extrinsics.insert(tx_hash);
+				});
+		}
+
+		log::debug!(target: LOG_TARGET,
+			"fatp::extrinsics_included_since_finalized {} from {} count: {} took:{:?}",
+			at,
+			recent_finalized_block,
+			all_extrinsics.len(),
+			start.elapsed()
+		);
+		all_extrinsics
+	}
+
+	/// For every watched transaction in the mempool registers a transaction listener in the view.
+	///
+	/// The transaction listener for a given view is also added to multi-view listener. This allows
+	/// to track aggreagated progress of the transaction within the transaction pool.
+	///
+	/// Function returns a list of currently watched transactions in the mempool.
+	async fn register_listeners(
+		&self,
+		view: &View<ChainApi>,
+	) -> Vec<(ExtrinsicHash<ChainApi>, Arc<TxInMemPool<ChainApi, Block>>)> {
+		log::debug!(
+			target: LOG_TARGET,
+			"register_listeners: {:?} xts:{:?} v:{}",
+			view.at,
+			self.mempool.unwatched_and_watched_count(),
+			self.active_views_count()
+		);
+		let view = Arc::from(view);
```

Comment:
Why you are creating here the `Arc` out of the view?

Reply:
Oh, ugly way to pass it to future. Improved in: fffc1fe05fd